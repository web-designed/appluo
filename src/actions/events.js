import uuid from 'uuid'
import moment from 'moment'
import database from '../firebase/firebase'


//*******************************************************
// EVENTS ACTION GENERATORS
//*******************************************************

export const addEvent = (event) => ({
   type:'ADD_EVENT',
   event
})

export const startAddEvent = (eventData = {}) => {
   return (dispatch) => {
      const {
         note = '', 
         cleaner = 'unknown', 
         cleanedAt = 0,
         createdAt = 0,
         place = ''
      } = eventData

      const event = { note, cleaner, cleanedAt, createdAt, place }

      return database.ref('events').push(event).then((ref) => { // add return to chain promises in the tests
         dispatch(addEvent({
            id: ref.key,
            ...event
         }))
      })
   }
}


export const removeEvent = (id) => ({
   type: 'REMOVE_EVENT',
   id
})

export const editEvent = (id, update) => ({
   type: 'EDIT_EVENT', 
   id,
   update
})

export const addComment = (eventId, { commentBody, commenter }) => ({
   type: 'ADD_COMMENT',
   eventId,
   comment: {
      id: uuid(),
      createdAt: moment().valueOf(),
      commenter,
      commentBody
   }
})

export const removeComment = (eventId, commentId) => ({
   type:'REMOVE_COMMENT',
   eventId,
   commentId
})

export const editComment = (eventId, commentId, update) => ({
   type:'EDIT_COMMENT',
   eventId,
   commentId,
   update
})