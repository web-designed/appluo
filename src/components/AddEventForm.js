import React from 'react'
import { connect } from 'react-redux'
import uuid from 'uuid'
import moment from 'moment'
import 'react-dates/initialize'
import { SingleDatePicker } from 'react-dates'
import { Link } from 'react-router-dom'

export class AddEventForm extends React.Component {

   state = {
      date: this.props.cleanedAt ? moment(this.props.cleanedAt) : moment(),
      focused: false,
      note: this.props.note ? this.props.note : '',
      place: this.props.place ? this.props.place : this.props.settings.places[0],
      error: '',
      cleaner: this.props.cleaner ? this.props.cleaner : this.props.settings.cleaners[0],
   }

   onCleanerChange = (e) => {
      const cleaner = e.target.value
      this.setState(() => ({ cleaner }))
   }

   onPlaceChange = (e) => {
      const place = e.target.value
      this.setState(() => ({ place }))
   }

   onNoteChange = (e) => {
      const note = e.target.value
      this.setState(() => ({ note }))
   }

   onDateChange = (date) => {
      this.setState(() => ({ date }))
   }

   onFocusChange = ({focused}) => {
      this.setState(() => ({focused}))
   }

   handleSubmit = (e) => {

      e.preventDefault()
      
      const event = {
         cleanedAt: this.state.date.valueOf(),
         createdAt: moment().valueOf(),
         cleaner: this.state.cleaner,
         place: this.state.place,
         note: this.state.note,
      }

      this.props.handleSubmit(event)
   }

   render(){

      console.log(this.props)

      return(
         <div>
            <form onSubmit={this.handleSubmit}>
               <div class="input-group mb-2">
                  <div class="input-group-prepend">
                     <label class="input-group-text" for="cleaner">Name</label>
                  </div>
                  <select 
                     class="custom-select" 
                     id="cleaner"
                     name="cleaner" 
                     onChange={this.onCleanerChange}
                     value={this.state.cleaner.toLowerCase()}
                  >
                     {
                        this.props.settings.cleaners.map( cleaner => {
                           return <option value={cleaner.toLowerCase()} key={cleaner.toLowerCase()}>{cleaner}</option>
                        })
                     }
                  </select>
               </div>
               <div class="input-group mb-2">
                  <div class="input-group-prepend">
                     <label class="input-group-text" for="place">Place</label>
                  </div>
                  <select 
                     class="custom-select" 
                     id="place"
                     name="place" 
                     onChange={this.onPlaceChange}
                     value={this.state.place.toLowerCase()}
                  >
                     {
                        this.props.settings.places.map( place => {
                           return <option value={place.toLowerCase()} key={place}>{place}</option>
                        })
                     }
                  </select>
               </div>
               <div class="input-group mb-2">
                  <div class="input-group-prepend">
                     <label class="input-group-text" for="date">Date</label>
                  </div>
                  <SingleDatePicker 
                     date={this.state.date}
                     onDateChange={this.onDateChange}
                     focused={this.state.focused}
                     onFocusChange={this.onFocusChange}
                     numberOfMonths={1}
                     isOutsideRange={() => false}
                  />
               </div>
               <div class="form-group">
                  <textarea 
                     class="form-control"
                     value={this.state.note}
                     name="note" 
                     placeholder="Add your note"
                     onChange={this.onNoteChange}
                  ></textarea>
               </div>
               <div class="form-group text-right">
                  <Link class="text-danger mr-2" to={'/'}>Cancel</Link>
                  <button class="btn btn-primary">Submit</button>
               </div>
            </form>
         </div>
      )
   }
}

const mapStateToProps = ({ settings }) => ({
   settings: settings
})

export default connect(mapStateToProps)(AddEventForm)