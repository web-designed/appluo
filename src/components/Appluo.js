import React from 'react'
import EventsList from './EventsList'

class Appluo extends React.Component {
   render() {
      return (
         <div>
            <EventsList />
         </div>
      )
   }
}

export default Appluo