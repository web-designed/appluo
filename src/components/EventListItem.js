import React from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import moment from 'moment'
import CommentCount from './CommentCount'

const EventListItem = ({ cleaner, place, cleanedAt, id, comments }) => {

   return (
      <div>
         <Link to={`/view/${id}`}>
            <div>
               <p>{place} | <strong>{cleaner}</strong> | {moment(cleanedAt).format('YYYY-MM-D')} | <CommentCount commentsCount={comments.length}/></p>
            </div>
         </Link>
      </div>
   )
}

export default connect()(EventListItem)