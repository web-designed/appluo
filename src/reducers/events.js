//*******************************************************
// EVENTS REDUCERS
//*******************************************************

   const eventsReducerDefaultState = []
   const eventsReducer = (state = eventsReducerDefaultState, action) => {
      switch(action.type){
         case 'ADD_EVENT':
            return state.concat(action.event)
         case 'REMOVE_EVENT':
            return state.filter(({ id }) => id !== action.id)
         case 'EDIT_EVENT':
            return state.map((event) => {
               if(event.id === action.id) {
                  return {
                     ...event,
                     ...action.update
                  }
               } else {
                  return event
               }
            })
         case 'ADD_COMMENT':
            return state.map((event) => {
               if(event.id === action.eventId){
                  return {
                     ...event,
                     comments: [...event.comments, action.comment]
                  }
               } else {
                  return event
               }
            })
         case 'REMOVE_COMMENT':
            return state.map((event) => {
               if(event.id === action.eventId){
                  return {
                     ...event,
                     comments: event.comments.filter((comment) => {
                        return comment.id !== action.commentId
                     })
                  }
               } else {
                  return event
               }
            })
         case 'EDIT_COMMENT':
            return state.map((event) => {
               if(event.id === action.eventId){
                  return {
                     ...event,
                     comments: event.comments.map((comment) => {
                        if(comment.id === action.commentId){
                           return {
                              ...comment,
                              ...action.update
                           }
                        } else {
                           return comment
                        }
                     }) 
                  }
               } else {
                  return event
               }
            })
         default:
            return state
      }
   }

export default eventsReducer