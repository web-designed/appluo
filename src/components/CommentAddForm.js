import React from 'react'
import { connect } from 'react-redux'
import uuid from 'uuid'
import { addComment } from '../actions/events'
import moment from 'moment'

export class CommentAddForm extends React.Component {

   state = {
      error: '',
      commentBody: '',
      commenter: 'kylu'
   }

   onCommentBodyChange = (e) => {
      const commentBody = e.target.value
      this.setState(() => ({ commentBody }))
   }

   handleSubmitComment = (e) => {

      e.preventDefault()

      if(!this.state.commentBody){
         this.setState(() => ({
            error: this.props.notifications.validate
         }))
      } else {
         const newComment = {
            commentBody: this.state.commentBody,
            commenter: this.state.commenter
         }
         //dispatch the action
         this.props.addComment(this.props.currentEventId, newComment)

         //clear the comment and the validation error
         this.setState(() => ({
            commentBody: '',
            error: ''
         }))
      }
   }

   render(){
      return (
         <div>
            <h3>Add comment</h3>
            <form onSubmit={this.handleSubmitComment}>
               <textarea 
                  name="commentBody"
                  value={this.state.commentBody}
                  onChange={this.onCommentBodyChange}
                  placeholder="Add your comment"
               ></textarea>
               {this.state.error && <p><span>{this.state.error}</span></p>}
               <p><button>Add Comment</button></p>
            </form>
         </div>
      )
   }
}

const mapDispatchToProps = dispatch => ({
   addComment: (id, newComment) => dispatch(addComment(id, newComment))
})

const mapStateToProps = state => ({
   notifications: state.settings.notifications.comment
})

export default connect(mapStateToProps, mapDispatchToProps)(CommentAddForm)