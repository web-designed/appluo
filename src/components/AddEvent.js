import React from 'react'
import AddEventForm from './AddEventForm'

const AddEvent = () => {
   return(
      <div>
         <h1>Add a cleaning event</h1>
         <AddEventForm />
      </div>
   )
}

export default AddEvent