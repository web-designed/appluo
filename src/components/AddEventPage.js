import React from 'react'
import AddEventForm from './AddEventForm'
import { addEvent } from '../actions/events'
import { connect } from 'react-redux'

const AddEventPage = (props) => (
   <div>
      <h1>Add a cleaning</h1>
      <AddEventForm 
         handleSubmit={ (event) => {
            props.onSubmit(event)
            props.history.push('/')
         }} 
         history={props.history} 
      />
   </div>
)

const mapDispatchToProps = (dispatch) => ({
   onSubmit: (expense) => dispatch(addEvent(expense))
})

export default connect(undefined, mapDispatchToProps)(AddEventPage)