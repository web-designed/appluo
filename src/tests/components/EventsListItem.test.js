import React from 'react'
import { shallow } from 'enzyme'
import { EventListItem } from '../../components/EventListItem'
import events from '../fixtures/events'

test('should render EventListItem correctly', () => {
   const wrapper = shallow(<EventListItem {...events[0]} />)
   expect(wrapper).toMatchSnapshot()
})