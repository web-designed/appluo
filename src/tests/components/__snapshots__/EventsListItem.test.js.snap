// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EventListItem correctly 1`] = `
<div>
  <Link
    replace={false}
    to="/view/0"
  >
    <div>
      <p>
        Bad
         | 
        <strong>
          Kylu
        </strong>
         | 
        1970-01-01
         | 
        <CommentCount
          commentsCount={2}
        />
         
        <span>
           | N
        </span>
      </p>
    </div>
  </Link>
</div>
`;
